src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.cs

```diff
@@ -260,5 +261,242 @@ public static bool HasAllSet(this BitArray bitArray)
 #else
         private static Regex CreateIntegerRegex() => new(IntegerRegexPattern, RegexOptions.Compiled, TimeSpan.FromMilliseconds(IntegerRegexTimeoutMs));
 #endif
+
+        /// <summary>
+        /// Compares two valid UTF-8 encoded JSON numbers for decimal equality.
+        /// </summary>
+        public static bool AreEqualJsonNumbers(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)
+        {
+            Debug.Assert(left.Length > 0 && right.Length > 0);
+
+            ParseNumber(left,
+                out bool leftIsNegative,
+                out ReadOnlySpan<byte> leftIntegral,
+                out ReadOnlySpan<byte> leftFractional,
+                out int leftExponent);
+
+            ParseNumber(right,
+                out bool rightIsNegative,
+                out ReadOnlySpan<byte> rightIntegral,
+                out ReadOnlySpan<byte> rightFractional,
+                out int rightExponent);
+
+            int nDigits;
+            if (leftIsNegative != rightIsNegative ||
+                leftExponent != rightExponent ||
+                (nDigits = (leftIntegral.Length + leftFractional.Length)) !=
+                            rightIntegral.Length + rightFractional.Length)
+            {
+                return false;
+            }
+
+            if (leftIntegral.Length == rightIntegral.Length)
+            {
+                return leftIntegral.SequenceEqual(rightIntegral) &&
+                    leftFractional.SequenceEqual(rightFractional);
+            }
+
+            // There is differentiation in the integral and fractional lengths,
+            // concatenate both into singular buffers and compare them.
+            scoped Span<byte> leftDigits;
+            scoped Span<byte> rightDigits;
+            byte[]? rentedLeftBuffer;
+            byte[]? rentedRightBuffer;
+
+            if (nDigits <= JsonConstants.StackallocByteThreshold)
+            {
+                leftDigits = stackalloc byte[JsonConstants.StackallocByteThreshold];
+                rightDigits = stackalloc byte[JsonConstants.StackallocByteThreshold];
+                rentedLeftBuffer = rentedRightBuffer = null;
+            }
+            else
+            {
+                leftDigits = (rentedLeftBuffer = ArrayPool<byte>.Shared.Rent(nDigits));
+                rightDigits = (rentedRightBuffer = ArrayPool<byte>.Shared.Rent(nDigits));
```

Comment:
Presumably this path is super rare?

Reply:
Super rare is accurate. It requires significands exceeding 256 digits.